### ISSUE TITLE
[RFC] Homebrew-like Registry Design

### ISSUE THREAD

**Issue by RLRabinowitz** [opentofu/opentofu#741]
# Overview

This is a proposal for the stable registry design.

**Disclaimer: This is not a proposal by the core team, it's my personal proposal for this.**

The design is loosely based on Homebrew's approach to providing Formulae to anyone. Homebrew manages the Formulae under one centralize repository ([homebrew-core](https://github.com/Homebrew/homebrew-core)). New Formulae are created via PRs. Adding new versions to a Formula requires a PR as well, and it can be created using `brew bump-formula-pr`

The design here is similar - Have a centralized repository containing index files for every provider and module. The index files contain version info for the provider, and other necessary metadata. Creating a provider/module would simply mean to create such an index file for the provider/module. Adding a new version would require a PR, but we could add some means to simplify and automate the process.

The requirements are available in this issue [#258](https://github.com/opentofu/opentofu/issues/258)

# Technical Overview

The design consists of:

- **A github repository**, managing the "source of truth" of the Registry's providers/modules and their versions
  - Updates are mainly using pull requests to add new providers/modules, or to add new versions to existing ones. Adds automation to simplify the process and to not force all provider/module authors to constantly publish versions manually
- **A serving layer** - Serves the provider/module versions and metadata 
- **A documentation website layer** - WIP

## The provider/module github repository

A repository containing a file per module and per provider. The file is a JSON file, containing all necessary information for OpenTofu to get the provider/modules versions and artifacts

### The provider/module JSON file

Will be of the following format:

```
{
  "keys" ["<PUBLIC_KEY_1>", "<PUBLIC_KEY_2>"],
  "versions": [{
    "version": "v3.0.352",
    "protocols": ["4.2", "5.0"],
    "shasum_url": "https://example.com/download/v3.0.352/SHASUM",
    "shasum_sig_url": "https://example.com/download/v3.0.352/SHASUM.sig",

    "targets": [{
      "os": "darwin",
      "arch": "amd64",
      "download_url": "https://example.com/download/v3.0.352/my-provider-darwin-amd64.zip",
      "file_name": "my-provider-darwin-amd64.zip",
      "shasum": "1"
    }, 
    {
      "os": "darwin",
      "arch": "arm64",
      "download_url": "https://example.com/download/v3.0.352/my-provider-darwin-arm64.zip",
      "file_name": "my-provider-darwin-amd64.zip",
      "shasum": "2"
    }]
  }]
}
```

### Adding a provider/module to the registry

Simply create a PR adding the JSON file for your provider/module, with all the versions and necessary metadata. After the PR is approved and merged, the file will be part of the main branch, and will later be served as part of the serving layer
We could also provide a script that could be used to generate such a file for your provider/module repository, and even create the PR with the change

### Adding a new version to an existing provider/module

Creating a PR adding the new version to the existing JSON file of the provider/module. We could add a bunch of methods to help automate the process, either to help out contributors with the version bump, or by completely automating the PR creation process

#### Auto-Approval of PRs

We could automatically approve and merge PRs of existing providers/modules. We could rely on the release in GitHub being the source-of-truth
We can simply auto-approve if the keys have not been changed in the PR, and the artifacts' signature match the public keys (But even that is not a must)

#### `tofu bump-provider-pr`

Similarly to brew's [brew bump-formula-pr](https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request#submit-a-new-version-of-an-existing-formula), we could create a command that updates the JSON based on your repository, forks the OpenTofu Registry repository and creates a PR updating it

#### A GitHub Action for use in the provider/module repository

We could create a GH action that creates the PR for you. Authors of provider/module repositories could add that GHA to their release workflow, and immediately publish an OpenTofu version

#### Periodic refresh of versions of existing providers/modules

Since we probably can't count on all provider/module authors to always bump the versions by themselves, we might need an automation around bumping the versions ourselves.
The automation should take GitHub API rate limits into account, and attempt to not hit those limits (or deal with them in some manner)

We run a cron GHA that runs `git ls-remote --tags` for all providers and modules, and compares them to the versions in the JSON files
For modules - Look for vX.X.X and X.X.X tags. Any tag that does not exist in the JSON files should be added - Use GH API to fetch the release and create the file
For providers - Look for vX.X.X tags. We can't solely rely on tags, not all tags have releases. So, in order to avoid making unnecessary GH API calls to those releaseless tags every time the GHA runs, we can do the following:
- Add a cached file in the GHA that lists how many times we've attempted to find a release for a tag, and it wasn't there. Once it goes over a specific threshold - Don't try using it again
- For every tag you find, before checking if it has a release, check the cached file. If we checked that tag too many times, don't check again. Otherwise - check if release exists
- If the release doesn't exist, update the number of attempts for that tag by 1
- The cached file should be enough, as losing it should not cause too many GH API calls (hopefully), but we can switch to having it stored in a more persistent manner (S3 or some DB) 

After the GHA is done, it'll create a PR adding those changes (or simply merge to main directly)

**An alternative option** is to not rely on tags, and attempt to only use the GH API for releases when checking for updates. This would require first-class handling of GH rate limits, similarly to as suggested in [the following RFC](https://github.com/opentofu/opentofu/issues/724) 

#### Initial filling of the registry's JSON files for existing providers/modules

We'd need to have the registry fill up with JSON file for all existing providers/modules. For that, we can run a GHA similar to the mentioned above to fill all of those, if we provide it with a set of providers and modules we'd want to populate.
It might take some time, as we'll probably get throttled by GitHub API along the way, but eventually all necessary JSON files will be created

## The serving layer

For the serving layer, we have a couple of options

### Option 1: GitHub Pages

- Easy to set up, simply serving the JSON files as they are
- Would require changes in the CLI - Instead of (or in addition to) supporting v1, we'd support this new JSON file format that has all the info we need to install providers/modules
- We shouldn't have any issues with bandwidth limitations, as GH Pages only has a [soft limit of 100GB per month](https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#usage-limits), but its 1GB hosting limitation might be a problem for expansion in the future, as we add more providers and modules to the repository
- ~~For this approach **we wouldn't support v1 API in the registry anymore** - meaning that legacy terraform and other tooling relying on the API won't be able to use the OpenTofu Registry~~ We could still support v1 API for the registry (minus the `X-Terraform-Get` header for module download) by rendering on a different branch that would contain the files in the necessary format for the v1 API (similarly to the mapping mentioned in Option 2)

### Option 2: CloudFront over replicated S3

- Similarly to the serving layer in [the following RFC](https://github.com/opentofu/opentofu/issues/724), this approach will have high-availability and expandability
- This would allow us to set up the hosted files in a different manner than how they are stored kept in the repository
  - Would require a pipeline after merge to main, to map the change to the format of the hosted files, and upload them to S3
  - Would allow us to keep the v1 API if we'd like (would require a small change, having the module download API return the URL in the body and not in the header. CLI would be adjusted as well to support both)

## The documentation layer - WIP

Currently, this RFC assumes the documentation website would be a separate system
That system could use the data hosted in the registry to generate persistent data and APIs that could used for documentation. Then, the documentation website could be a similar effort to that mentioned in other RFCs, like https://github.com/opentofu/opentofu/issues/724 or https://github.com/opentofu/opentofu/issues/722

# Upsides

- Easy to host yourself as your own registry - Simply create static hosting of JSON files in the decided format
- Using GitHub Pages - we solely rely on GitHub for the availability of the registry (which we need to anyway)
- GitHub API rate limits are pretty much completely avoided, nullifying the need to work around them
- Updating providers/modules/versions, and the serving layer are fairly simple, leaving little room for error or data inconsistency at runtime
- A point that might not seem relevant to the community, but is very relevant: No complicated service for the core-team to maintain. Any time spent supporting the uptime and upkeep of a service means less time for the core team to handle issues/bugs/enhancement requests in OpenTofu

# Downsides

- Registering new signing keys, providers, and modules is not fully self-service.
- Creating new providers and modules require the approval of a core-team member

Any feedback would be appreciated 🙏🏻

**Comment by cube2222** [comment]
So, some feedback and thoughts from my side:

It would be good to have a section discussing all requirements and how this design addresses them, like [here](https://github.com/opentofu/opentofu/issues/724) (in the "Specific discussion of each requirement" section).

---

Regarding the Option 1 (serving using GitHub Pages) - there's no reason we couldn't maintain as compatible with the v1 registry as with option 2. GitHub Pages support serving from a separate branch, so instead of rendering to an S3 bucket, we'd just be rendering the static files to a separate branch (which would adhere to the schema).

---

> GitHub API rate limits are pretty much completely avoided, nullifying the need to work around them

For providers, as you described in the RFC, we'd still need to somewhat care about the rate limits, as they rely on releases.

---

Overall I think this is a cool design. I see the tradeoff as follows:

The main upside is that with this, the layer that requires high-availability (so really what is described in this RFC, the v1 registry protocol serving) is very simple and the maintenance effort is low. There is no database to manage and backup, etc. This is a very big upside.

Additionally, storing all info about each version in git gets us very good visibility and auditability.

The main downside imo is that I don't see a "simple" extension for the documentation+search functionality and for the download metrics collection. In other words, we'd need to maintain a separate service (a fairly simple one, sure) to handle the documentation serving page. Same for the metrics gathering, we'd need to host something and receive download events sent from the OpenTofu CLI.

In other words, even though only this part is mission-critical to be highly-available, we'd overall need to maintain 3 (fairly simple) services. While with the other design currently proposed for instance we only have one service handling all functionalities, but with more complexity in it (there are still multiple components to it, and it relies on a database).

**Comment by cube2222** [comment]
One more thing @RLRabinowitz, we'll need to handle the hashicorp=>opentofu namespace redirect, and I believe the easiest way to do that is introduce a new field to the provider/module index file. It would be called `repository` and indicate the repository that should be scraped.

Basically, we need to serve e.g. hashicorp/aws in a way that points to our github artifacts, and we want the scraper to scrape our repo. Fortunately the download URLs are absolute, so we just serve our releases download URLs. But I think it's worth including this note.

(We can also just hardcode it instead of making a configurable repository override, either way works, but I think it's worth including it into the RFC).

**Comment by cube2222** [comment]
Also, just so that this proposal is complete, and contains a proposal on how to support docs/search.

With this design it would basically be almost completely separate. I'd only reuse the index contained in the repo to power the search, but that's it. It's worth noting that the availability requirements here are much lower and moving parts are less "expensive" maintenance-wise.

Now, once you open a specific provider, you also get a selector to switch versions (with latest selected by default). For a given version, the frontend would make an API call to get all info needed to render the given provider+version specific info.

We’d have a cloudfront lambda here (or whatever similar construct) that would do the following:

1. Gets request
2. Checks if docs data for the provider version is in S3
3. If yes, return the data (and cache it on cloudfront), otherwise continue.
4. Check if there’s a special object in S3 called “<provider>-<version>-processing.lock” and it’s newer than 2 minutes. If there is, return a non-cached “still loading” (which the frontend renders as “give us a moment to fetch the documentation for you”). If not, continue
5. Create the above special object, kick off an async lambda to load the given provider version into S3, then return a non-cached “still loading”.

If the frontend receives “still loading” it will poll the backend every 5 seconds.

The async lambda will git clone the provider repo (avoiding rate limits thanks to git clone), checkout the given version tag, and then process the documentation into the format required by the frontend. Finally, it puts it into S3.

With this it’s basically just cloudfront + lambda + s3. Very similar design to our alpha registry, but avoids api rate limits. To be clear, this is a completely separate S3 bucket than the one serving the v1 registry api.

Basically, in order not to have a job queue that would schedule fetching the docs for each new indexed provider version, which might require some queue, or database, or whatever, we just do the approach where we index on request. And the lock is there to make sure that there’s only one simultaneous lambda trying to load docs for a given provider version (so multiple people fetching a single unindexed version at once don’t trigger multiple “load” lambdas)

Basically, this makes the search and docs a completely separate system from the serving registry, with the only common part being that it uses the registry repo as an “index” for search and listing purposes.

**Comment by kislerdm** [comment]
Hey Arel, thanks for the proposal.

I'd like to share a couple of thoughts.

1. Even though I support the proposed approach for explicit notification from provider builders about new release, I have a slight concern that the proposed release process for new version would negatively impact the adoption of OpenTofu. To illustrate:

> Creating a PR adding the new version to the existing JSON file

If I understood the quote right, new version's availability will depend on the reviewers bandwidth. I find this limiting, or imposing high costs to make sure that there is high capacity to review PR assuming that review requires manual labour and human approval. I understand that the process can be automated, but I feel that it would defeat the purpose of PR review, wouldn't it?

For context, when I, as provider/module developed, release a new version, I expect it to be available for my customers in minutes, not hours.

2. I find the proposal of explicit provider creation using GitHub more appropriate compared to the status quo used today. However, I'd like to propose a binary encoding protocol - e.g. AVRO, or protobuf, for two reasons:
- Ease of schema maintenance in a BW and FW compatible fashion;
- Economic, JSON file's size will grow quite big quite fast which would impose avoidable limitations on bandwidth.

WDYT?

**Comment by cube2222** [comment]
> Even though I support the proposed approach for explicit notification from provider builders about new release, I have a slight concern that the proposed release process for new version would negatively impact the adoption of OpenTofu.

@kislerdm I believe the default way of refreshing would be the periodic github action described under `Periodic refresh of versions of existing providers/modules`. This doesn't require any action from the author, but can have a delay.

If the author wants versions to be published immediately, they can do so via a PR. But that would be an extension after we have the "default way" working.

There is unfortunately no way we can have both - short delay, and no action required on the authors side. So the above let's the author choose. If they don't do anything, it will be indexed with a delay. If they implement the action in their repo, it will be published right away.

> I'd like to propose a binary encoding protocol 

I get the pros, but I believe that with the registry size (20k providers/modules with 10-20 versions on average) the file size gains are irrelevant. At the same time, binary protocols aren't easy to read, and aren't easy to review in pull requests. Thus, I'm very strongly in favour of storing it as JSON files in the repository.

Besides, protobuf doesn't really give you any magic for backward-forward compatibility afaik. Sure, it does as far as binary formats go (because each field gets a unique index), but JSON already satisfies this by having separate fully-written-out keys for each field. Also worth keeping in mind that the json format is an implementation detail and OpenTofu will consume a mostly-V1-registry-protocol-compatible API that is rendered based on those JSON files.

**Comment by kislerdm** [comment]
@cube2222 Thanks for your reply Kuba. 

> I believe the default way of refreshing would be the periodic github

Did I get this point right, PR would be only required to create a new provider or module, the registry will track releases on a periodic basis.

About JSON, you are right, if the readability is critical, binary encoding is not convenient at all. Although, we shall keep in mind auto-validation/linting because manual reading through JSON with 100+ LoC is quite error prone.

Thanks!

**Comment by cube2222** [comment]
@kislerdm 

> Did I get this point right, PR would be only required to create a new provider or module, the registry will track releases on a periodic basis.

Precisely! We'd also do an initial backfill, so that the most popular providers/modules (or maybe all of the existing one?) are accessible right away.

> Although, we shall keep in mind auto-validation/linting because manual reading through JSON with 100+ LoC is quite error prone.

I think that's an excellent idea! Maybe we could use something like json-schema down the line?

**Comment by MikeMcQuaid** [comment]
Hello, Homebrew project leader here 👋🏻. I got linked to this (cool!) idea and figured some of my experience from working on Homebrew for the last 14 years might be useful here!

> * **A github repository**, managing the "source of truth" of the Registry's providers/modules and their versions

You probably want to be quite careful with how you design the structure of this repository. We went through a "sharding" process recently so that we have formulae/casks in our largest repositories split into subdirectories (https://github.com/Homebrew/homebrew-core/tree/master/Formula https://github.com/Homebrew/homebrew-cask/tree/master/Casks).

The `git` performance is _dramatically_ better in this format. `git` does not like having deep histories combined with too many files or folders in a directory.

My understanding of the above is that end-users will not cloning this repository other than to contribute to the project so the `git clone`/`git fetch` speeds are probably less important to you but they are also heavily impacted by this.

---

Using JSON files rather than a more involved format (e.g. YAML) seems very sensible. The security implications of these files being purely simple, structured data are much nicer.

---

Nothing else jumps out here but: I'm happy to answer any more questions if there's anything relevant I can add.

Regardless: good luck and hi from Homebrew land 👋🏻 ❤️ 🍻

**Comment by cube2222** [comment]
Thanks a lot for the input @MikeMcQuaid, appreciate it a lot and we'll definitely reach out for advice, if we have any questions!

**Comment by kislerdm** [comment]
Hey folks! I prepared architecture design diagrams to facilitate RFC "digestion". Please let me know if it makes sense, so we could think of such diagrams as part of RFC guidelines. Thanks!

![registry-l2-rfc741-list-versions](https://github.com/opentofu/opentofu/assets/13434797/02e3bd7a-81ab-40b9-98be-270118f5f6ab)
Source: [plantuml link](http://www.plantuml.com/plantuml/uml/VP5Hgvj048NV_HLRIuc1cuP2e2-bND3qff3q1ateGobXpOvsODsHtJNKVz_HQ6AWhK_kp9blpvbzTbvQtrGwU8Fch1i5ubEyMiQhKqp6IpHWFmU1Hwz1_4Y_n6y_hZxy5BUUK2XBAoj1XNYjmHoeQ8I54fsthN1WBsXAKLkwe0BhbXMfHeCJ5puW6HSKH9unuMru8TufwI6QjVmjThk5KhDv60HxdY0JDKyYF38td8yLnPK4Bg2fxij1rZeFrMdzsqTdZKyJThpz6l9v9uqimTrj2McK6ArfEakjs_UCkiMCwcw0HxCXqLQsO3lM2_fDaujzrmy70KCgy99Zo4UAXPfO4tD3KpaPJxdXq9DjUmR__DFUb3n9MgB_rUIFn6KA6gIx6jkGytVcp8apASYvszDpDD8YHvpSnHe51HhqIAQxBc3UAUcUuBhNOuUU3zM-DAuf1QdSVvvT59QgtjNXiCy6nlRA698CaBygaqXDT0bexC32Kx-zR3F8qc7Xewi4Qatjg7i4ZhJXYR8chz3Vj_t03hBDw_VJTlso_fP4V7m6evfA_m40).

![registry-l2-rfc741-list-versions-alt](https://github.com/opentofu/opentofu/assets/13434797/caa24e1a-fe4f-4efb-9341-ffd106734c19)
Source: [plantuml link](http://www.plantuml.com/plantuml/uml/bP9HYnf13CVVzrEaMueUQALKw4ifLrRlB6ZlS2ttK8gCJjm6PYSoaxNTRz-uwghqUb2VJFBFB_zCzZQA2LALBdb5VkqgY_0n6mwouJ9ZBuOyXaz98YGEuVlyBdlzOVZk1xGri6IAO4hW3JniqIzuKq70WgA46YA67Va2je5tP374GScsSXXXfm6nZpqmJZ1u8xJ3PCEi0QGjgGzlh17iTchzzMUplhMTcpH97hMTVRVgGVfDXwGtfupLp0Xtw7ZRf9EyZeBbSlnRkcj7lGjsDlsIQZmptXGOpnx1U0idduxNnhdwhQBQdRl7jRGvFouuDRMEqyJdfnmonvMz2ofFZn85bIX6Lp7Nf81RLasc1SV5NdvFCgbMCDT29E5GDmpzyq-BpvEZjkXUsZnXWMDqSfY__o_gnHiM93-hrRNFmHmTceWnRQP5EJCx4TPiSQNLXhuYRm9Xx5tCKet53NaIOhyVboXlETeVLozsk73Kz27RfCu7cAEnyONl2pQ1oyRNOl6O7odJ0wNTfO4yfzHvza8tGaTxC0WtxvTFMjdXaET7psk_Xb_Q-3UnDT-gNpHmcjvCK72rAadEzxdoakIJXwVbT7m__ZhgQdYBtbQb-mC0).

**Comment by kislerdm** [comment]
@RLRabinowitz Hey Arel! Could you please add a couple examples how to use it to the readme? Thanks!

**Comment by cube2222** [comment]
Awesome PoC @RLRabinowitz!

If we decide to go with this RFC, we'll probably want to implement sharding as MikeMcQuaid suggested, cause the performance of the repository in the GitHub UI is already atrocious.

> You probably want to be quite careful with how you design the structure of this repository. We went through a "sharding" process recently so that we have formulae/casks in our largest repositories split into subdirectories (https://github.com/Homebrew/homebrew-core/tree/master/Formula https://github.com/Homebrew/homebrew-cask/tree/master/Casks).

So basically split the providers/modules directory into a,b,c,d,e,f,g, etc. directories, which indicate the first letter of the namespace. And then the actual namespaces listed inside of that.

**Comment by cube2222** [comment]
Note: this design has been chosen for implementation.

More details in today's (2023-11-03) update in https://github.com/opentofu/opentofu/issues/258

**Comment by mering** [comment]
What is the purpose of the `shasum_url` if the `shasum` is explicitly included for each `target`?

**Comment by Yantrio** [comment]
> What is the purpose of the shasum_url if the shasum is explicitly included for each target?

Both values are returned in the v1 api. and are consumed by `tofu` for validation here: https://github.com/opentofu/opentofu/blob/c8acedd88512eae371d3131b380238a7dec42010/internal/getproviders/registry_client.go#L305-L328

If my memory serves me correctly, the checksum is used to valiate the download is not corrupted, and then the shasums url and signature url are used to validate the signatures

**Comment by mering** [comment]
> If my memory serves me correctly, the checksum is used to validate the download is not corrupted, and then the shasums url and signature url are used to validate the signatures

What is the purpose of checking the signature of the shasums file if it is never checked if the target shasum is actually the same as the one declared in that file?

**Comment by cube2222** [comment]
@mering @Yantrio to be more specific, it works as follows:

It fetches both the registry-stored checksum and fetches the checksum file
It finds the relevant file in the checksum file and makes sure that the checksum there is equal to the checksum in the registry
Then it checks the signature, which is a signature **of the whole checksum file**
Finally, it downloads the artifact and uses the checksum to verify it

Overall, this means that:
- The downloaded artifact is correct (checksum matches)
- It's authorized by the key (there's a signature to check)
- We only have a single signature for all downloads (since we sign the whole shasums file, not each checksum on its own)
- If we want, it can prove that the release hasn't been tampered with. That is, once its scraped and added to the registry, the registry now has a checksum, so it will be able to guarantee that the binaries / checksums haven't changed.

But the existence of the checksums file itself basically means that you only need to have a single signature.

**Comment by mering** [comment]
Shouldn't it be enough to do the `shasum` in `shasum_url` and `shasum_sig_url` matching checks on package addition (e.g. PR) such that the client only needs to check the checksum of the downloaded file?

The client needs to trust GitHub anyways, so where is the additional security if the client does these additional downloads and checks?
If you control GitHub you could equally well modify the public signing key, the `shasum_url` and `shasum_sig_url` fields...

**Comment by cube2222** [comment]
@mering while controlling github "in general" will obviously let you compromise the registry (where only tofu lockfiles will help you) this does limit what an attacker can do if they only control the provider author's account.

There's also the point of simply being as compatible with the existing protocol as possible, and making as few changes prior to the stable release as possible.
